### DISCUSSION TITLE
Keda Selenium scaler: scalehandler	Error getting scale decision context deadline exceeded (Client.Timeout exceeded while awaiting headers)

### DISCUSSION THREAD

**Discussion by deepakguna** [kedacore/keda#3649] [Q&A / Need Help]
Keda Selenium scaler: scalehandler	Error getting scale decision context deadline exceeded (Client.Timeout exceeded while awaiting headers)
I am trying to use keda selenium scaler for my grid in AWS k8s - however it is having issue with keda-operator and metrics server.

Any help much appreciated. 

Following is my keda scaledobject definition: 
`apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: selenium-grid-chrome-scaledobject
  namespace: selenium
  labels:
    deploymentName: selenium-chrome-node-deployment
spec:
  minReplicaCount: 0
  maxReplicaCount: 30
  scaleTargetRef:
    name: selenium-chrome-node-deployment
  triggers:
    - type: selenium-grid
      metadata:
        url: 'https://internal.selenium.pdosandbox.aws.test.com/graphql'
        browserName: 'chrome'
        unsafeSsl : 'true'`

When it comes to keda operator  i see the following stack trace. 

1.6625424133568501e+09	INFO	controller.scaledobject	Reconciling ScaledObject	{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.6625431543361835e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
1.6625431843325372e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
1.6625432143320718e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
1.6625432443332555e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
1.662543274334177e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
1.6625433043334424e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}

**Comment by Wolfe1** [comment]
Seems as though keda can not reach your graphql endpoint for some reason. I would ssh into the keda operator and run a ping to the base url to see if it can be reached by the operator:

**ping internal.selenium.pdosandbox.aws.test.com**

**Comment by deepakguna** [comment]
Thank you @Wolfe1  for a qick response. How do we ssh in to Keda operator? it doesn have /bin/sh or /bin/bash

**Comment by Wolfe1** [comment]
Ah you are right, you could run a busybox pod from the same namespace and effectively do the same thing. I think the issue lies in that URL being able to resolve within the namespace the keda operator is in (default I believe is just "keda").

**Reply by deepakguna** [reply]
Do you recommend to have Keda in the same namespace as the selenium grid's deployements?

**Reply by Wolfe1** [reply]
I don't think that should be necessary. You need to just get your networking in order to allow keda to see the grid and scale accordingly.

Perhaps looking into something like the first answer here would help: https://stackoverflow.com/questions/37221483/service-located-in-another-namespace

**Reply by deepakguna** [reply]
The curl to the selenium GraphQL end point is working fine from an alpine container in keda namespace . So it is able to talk to the grid. Not  sure what else is the problem

curl -X POST -H "Content-Type: application/json" --data '{"query": "{ grid {
maxSession, sessionCount } }"}' -s https://internal.selenium.pdosandbox.aws.test/graphql

{
  "data": {
    "grid": {
      "maxSession": 2,
      "sessionCount": 0
    }
  }
}

**Reply by deepakguna** [reply]
Let me clarify few more things, I have installed Keda in default "keda" namespace and the scaled object for this chrome browser in selenium namespace where all my grid related deployments are . 

I couldn't  figure out what could be the reason for the below mentioned error in keda operator : 

ternal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
1.6629726497229156e+09	ERROR	scalehandler	Error getting scale decision	{"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
	/workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
	/workspace/pkg/scaling/scale_handler.go:149
dgunasek@C02F548PMD6M ~ % 9325942829332e+09     INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.6629707727643526e+09  INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.6629715167609472e+09  INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.6629715318505926e+09  INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.662971533724117e+09   INFO    scaleexecutor   Successfully set ScaleTarget replicas count to ScaledObject minReplicaCount     {"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "Original Replicas Count": 1, "New Replicas Count": 0}
1.6629715467895696e+09  INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.662971726840714e+09   INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.6629717418247926e+09  INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.662971743730643e+09   INFO    scaleexecutor   Successfully set ScaleTarget replicas count to ScaledObject minReplicaCount     {"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "Original Replicas Count": 1, "New Replicas Count": 0}
1.6629717567875538e+09  INFO    controller.scaledobject Reconciling ScaledObject{"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "selenium-grid-chrome-scaledobject", "namespace": "selenium"}
1.6629725897225997e+09  ERROR   scalehandler    Error getting scale decision    {"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
        /workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
        /workspace/pkg/scaling/scale_handler.go:149
1.662972619721835e+09   ERROR   scalehandler    Error getting scale decision    {"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
        /workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
        /workspace/pkg/scaling/scale_handler.go:149
1.6629726497229156e+09  ERROR   scalehandler    Error getting scale decision    {"scaledobject.Name": "selenium-grid-chrome-scaledobject", "scaledObject.Namespace": "selenium", "scaleTarget.Name": "selenium-chrome-node-deployment", "error": "Post \"https://internal.selenium.pdosandbox.aws.test.com/graphql\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).checkScalers
        /workspace/pkg/scaling/scale_handler.go:278
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop
        /workspace/pkg/scaling/scale_handler.go:149

**Reply by deepakguna** [reply]
Scaled object status: 

> 
> NAME                                SCALETARGETKIND      SCALETARGETNAME                   MIN   MAX   TRIGGERS        AUTHENTICATION   READY   ACTIVE   FALLBACK   AGE
> 
> selenium-grid-chrome-scaledobject   apps/v1.Deployment   selenium-chrome-node-deployment   0     30    selenium-grid                    True    False    False      5d1h

**Reply by deepakguna** [reply]
@Wolfe1  please chime in your thoughts

**Reply by Wolfe1** [reply]
@deepakguna I really haven't a clue, I am just a user of keda with some experience utilizing it to scale selenium grid. I would suggest trying to bring this to some KEDA developers attention in the [slack channel](https://github.com/kedacore/keda#community) or create an issue (rather than a discussion) in case this is a bug.

**Reply by deepakguna** [reply]
Thank you @Wolfe1 , i will reach out to them.

**Comment by deepakguna** [comment]
@zroubalik @prashanth-volvocars @arschles your help much appreciated.

**Reply by prashanth-volvocars** [reply]
> @zroubalik @prashanth-volvocars @arschles your help much appreciated.

Hey @deepakguna 

There should be a similar error event when scaled object was created. Failed to connect to <your url> ….. If you see it then its mostly that keda is not able to reach you hub under the given url.

i also suggest you try doing what @Wolfe1 had recommended above. Sshing into any pod in keda namespace and try to see if you can connect to the graphql url. If yes then we can look into it further.

**Reply by deepakguna** [reply]
Thank you @prashanth-volvocars  for the quick reply . I am able to do a curl from keda namespace from within an alpine container 
curl -X POST -H "Content-Type: application/json" --data '{"query": "{ grid { maxSession, sessionCount } }"}' -s https://internal.selenium.pdosandbox.aws.test.com/graphql

**Reply by deepakguna** [reply]
I also see errors in metrics server.  which i have listed in https://github.com/kedacore/keda/discussions/3647

So basically , though i am able to curl to the graphql endpoint from the keda's namespace, kedaoperator is having difficulty to do so. I am clueless on this for quite some time. your insight on this is much appreciated.

**Reply by deepakguna** [reply]
Here is the scaledobject:  i dont see any error  @prashanth-volvocars 

k describe hpa keda-hpa-selenium-grid-chrome-scaledobject -n selenium

```
Name:                                               keda-hpa-selenium-grid-chrome-scaledobject
Namespace:                                          selenium
Labels:                                             app.kubernetes.io/managed-by=keda-operator
                                                    app.kubernetes.io/name=keda-hpa-selenium-grid-chrome-scaledobject
                                                    app.kubernetes.io/part-of=selenium-grid-chrome-scaledobject
                                                    app.kubernetes.io/version=2.7.1
                                                    deploymentName=selenium-chrome-node-deployment
                                                    scaledobject.keda.sh/name=selenium-grid-chrome-scaledobject
Annotations:                                        <none>
CreationTimestamp:                                  Wed, 07 Sep 2022 14:49:58 +0530
Reference:                                          Deployment/selenium-chrome-node-deployment
Metrics:                                            ( current / target )
  "s0-seleniumgrid-chrome" (target average value):  <unknown> / 1
Min replicas:                                       1
Max replicas:                                       30
Deployment pods:                                    0 current / 0 desired
Conditions:
  Type            Status  Reason              Message
  ----            ------  ------              -------
  AbleToScale     True    SucceededGetScale   the HPA controller was able to get the target's current scale
  ScalingActive   False   ScalingDisabled     scaling is disabled since the replica count of the target is zero
  ScalingLimited  False   DesiredWithinRange  the desired count is within the acceptable range
Events:           <none>


```

**Reply by deepakguna** [reply]
@prashanth-volvocars @Wolfe1  finally i narrowed down the issue . The issue was the size of the response of the following request from selenium scaler: it returns huge volume of data - around 3 MB of data and this results in time out. When i reduced the test size to 3 or less or it needs increase in   pollingInterval up to 360 seconds. even then 20% of tests fails due to unavailability of pods 



`
func (s *seleniumGridScaler) getSessionsCount(ctx context.Context, logger logr.Logger) (int64, error) {
	body, err := json.Marshal(map[string]string{
		**"query": "{ grid { maxSession, nodeCount }, sessionsInfo { sessionQueueRequests, sessions { id, capabilities, nodeId } } }",**
	})`

**Reply by deepakguna** [reply]
@prashanth-volvocars  the file is too large at 109 MB when i attempted to run 60 tests in 30 parallel threads. PFA in g-drive : 
https://drive.google.com/file/d/1sE0eLWsyvplLuZaVPVd1lx7qWVwCE5_V/view?usp=sharing



If we look at the router logs it is unable to process the concurrent requests as each one has to serve 100+ MB 

System info: host: 'selenium-router-deployment-946949964-q5swx', ip: '100.122.15.200', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.196-108.356.amzn2.x86_64', java.version: '11.0.16'
Driver info: driver.version: unknown
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler.makeCall(NettyHttpHandler.java:65)
	at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
	at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:56)
	at org.openqa.selenium.remote.http.netty.NettyHttpHandler.execute(NettyHttpHandler.java:49)
	at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
	at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:56)
	at org.openqa.selenium.remote.http.netty.NettyClient.execute(NettyClient.java:98)
	at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:55)
	at org.openqa.selenium.grid.security.AddSecretFilter.lambda$apply$0(AddSecretFilter.java:40)
	at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:56)

**Reply by deepakguna** [reply]
I believethe extension in the sessionQueueRequests is causing the file to bloat.
